packages/flutter/lib/src/widgets/autofill.dart

```diff
@@ -0,0 +1,182 @@
+///          },
+///        ),
+///        // Again the address fields are grouped together for the same reason.
+///        if (!isSameAddress) AutofillGroup(
+///          child: Column(
+///            children: <Widget>[
+///              TextField(
+///                controller: billingAddress1,
+///                autofillHints: <String>[AutofillHints.streetAddressLine1],
+///              ),
+///              TextField(
+///                controller: billingAddress2,
+///                autofillHints: <String>[AutofillHints.streetAddressLine2],
+///              ),
+///            ],
+///          ),
+///        ),
+///        const Text('Credit Card Information'),
+///        // The credit card number and the security code are grouped together as
+///        // some platforms are capable of autofilling both fields.
+///        AutofillGroup(
+///          child: Column(
+///            children: <Widget>[
+///              TextField(
+///                controller: creditCardNumber,
+///                autofillHints: <String>[AutofillHints.creditCardNumber],
+///              ),
+///              TextField(
+///                controller: creditCardSecurityCode,
+///                autofillHints: <String>[AutofillHints.creditCardSecurityCode],
+///              ),
+///            ],
+///          ),
+///        ),
+///        const Text('Contact Phone Number'),
+///        // The phone number field can still be autofilled despite lacking an
+///        // `AutofillScope`.
+///        TextField(
+///          controller: phoneNumber,
+///          autofillHints: <String>[AutofillHints.telephoneNumber],
+///        ),
+///      ],
+///    );
+///  }
+/// ```
+/// {@end-tool}
+class AutofillGroup extends StatefulWidget {
+  /// Creates a scope for autofillable input fields.
+  ///
+  /// The [child] argument must not be null.
+  const AutofillGroup({
+    Key key,
+    @required this.child,
+  }) : assert(child != null),
+       super(key: key);
+
+  /// Returns the closest [AutofillScope] which encloses the given context.
+  ///
+  /// In order to interact with the platform's autofill mechanism,
+  /// [AutofillTrigger]s need to call [AutofillScope.attach] on their closest
+  /// [AutofillScope], instead of calling [TextInputClient.attach].
+  static _AutofillGroupState of(BuildContext context) {
+    final _AutofillScope scope = context.dependOnInheritedWidgetOfExactType<_AutofillScope>();
+    return scope?._scope;
+  }
+
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  @override
+  _AutofillGroupState createState() => _AutofillGroupState();
+}
+
+class _AutofillGroupState extends State<AutofillGroup> with AutofillScopeMixin {
+  @override
+  Iterable<AutofillClient> get autofillClients {
+    final List<AutofillClient> clients = <AutofillClient>[];
+    void visit(Element element) {
+      if (element is AutofillScope)
+        return;
+      if (element is AutofillClient) {
```

Comment:
In what usage case are the Element itself the AutofillClient or AutofillScope?

Reply:
For instance, to implement their own custom input widget, the developer extended `SingleChildRenderObjectWidget` and created their own `Element` subclass for it, instead of subclassing a `StatefulWidget`?

Reply:
Why would you want to do that? I am wondering what you can do with that that isn't possible with just supporting this for StatefulWidgets to justify making this more complex.

Reply:
That makes it a little bit easier to migrate existing code bases that implement `TextInputClient` using an `Element` instead of `State` I think?

Reply:
Is that something people do? I am not convinced that's worth supporting.